src/anyio/_core/_synchronization.py

```diff
@@ -792,41 +813,72 @@ def from_max_per_second(cls, max_per_second: int, /) -> Self:
         """
         return cls(1, 1 / max_per_second)
 
+    def _update_available(self, slot_time: float | None = None) -> bool:
+        """
+        Update the number of available tokens.
+        Returns True if tokens are available.
+
+        :param slot_time: If set, do not update if the next-slot time has
+                          changed (via a call to statistics form another task)
+
+        :return: a flag whether a slot is available
+        """
+        now = current_time()
+
+        if self._next is None:
+            # First call.
+            self._next = now + self._interval
+            return self._available > 0
+
+        if slot_time is None:
+            if self._next > now:
+                # Need to sleep?
+                return self._available > 0
+        elif slot_time != self._next:
+            # Another task called .statistics() while we were about to wake up
+            assert self._available > 0
+            return True
+
+        # Calculate burst increment
+        if self._burst_rate:
+            tokens = int(
+                self._burst_rate
+                * (self._available + self._tokens * (now - self._next) / self._interval)
+            )
+            if tokens < 0:
+                tokens = 0
+            elif self._burst_max is not None:
+                tokens = min(self._burst_max, tokens)
+        else:
+            tokens = 0
+        self._available = self._tokens + tokens
+        self._next = now + self._interval
+        return True
+
     async def acquire(self) -> None:
         """
         Ensure that the rate limit is not exceeded.
 
         This method must be called before running the related rate-sensitive operation.
-
         """
         async with self._lock:
             # Initialize or refresh the next token pool refresh time
-            now = current_time()
-            if self._next is None or self._next <= now:
-                self._next = now + self._interval
-
-            # If no tokens are available, wait until the next token pool refresh time
-            if self._available == 0:
-                await sleep_until(self._next)
-                self._available = self._tokens
+            if not self._update_available():
+                slot_time = self._next
+                await sleep_until(cast(float, self._next))
+                self._update_available(slot_time)
 
             self._available -= 1
 
     def statistics(self) -> RateLimiterStatistics:
-        """Return the statistics for the underlying semaphore."""
-        if self._next is None:
-            return RateLimiterStatistics(
-                tasks_waiting=0,
-                available_tokens=self._available,
-            )
+        """Return the statistics for the underlying lock."""
 
-        if current_time() >= self._next:
-            self._available = self._tokens
-            self._next = None
+        # Refresh token availability
+        self._update_available()
```

Comment:
I don't like `statistics()` affecting the internal state of the limiter. I did away with that in my own recent changes to the rate limiter PR.